src/lib/collision_prevention/CollisionPrevention.cpp

```diff
@@ -376,180 +509,148 @@ CollisionPrevention::_sensorOrientationToYawOffset(const distance_sensor_s &dist
+		const Vector2f &setpoint_vel,
+		const hrt_abstime now, float &vel_comp_accel, Vector2f &vel_comp_accel_dir)
+{
+	for (int i = 0; i < BIN_COUNT; i++) {
+		const float max_range = _data_maxranges[i] * 0.01f;
 
-						if (distance < max_range) {
-							delay_distance += curr_vel_parallel * (data_age * 1e-6f);
-						}
+		// get the vector pointing into the direction of current bin
+		float bin_angle = wrap_2pi(vehicle_yaw_angle_rad
+					   + math::radians((float)i * BIN_SIZE + _obstacle_map_body_frame.angle_offset));
 
-						const float stop_distance = math::max(0.f, distance - min_dist_to_keep - delay_distance);
-						const float vel_max_posctrl = xy_p * stop_distance;
+		const Vector2f bin_direction = { cosf(bin_angle), sinf(bin_angle) };
+		float bin_distance = _obstacle_map_body_frame.distances[i];
 
-						const float vel_max_smooth = math::trajectory::computeMaxSpeedFromDistance(max_jerk, max_accel, stop_distance, 0.f);
-						const float projection = bin_direction.dot(setpoint_dir);
-						float vel_max_bin = vel_max;
+		// only consider bins which are between min and max values
+		if (bin_distance > _obstacle_map_body_frame.min_distance && bin_distance < UINT16_MAX) {
+			const float distance = bin_distance * 0.01f;
 
-						if (projection > 0.01f) {
-							vel_max_bin = math::min(vel_max_posctrl, vel_max_smooth) / projection;
-						}
+			// Assume current velocity is sufficiently close to the setpoint velocity, this breaks down if flying high
+			// acceleration maneuvers
+			const float curr_vel_parallel = math::max(0.f, setpoint_vel.dot(bin_direction));
+			float delay_distance = curr_vel_parallel * _param_cp_delay.get();
 
-						// constrain the velocity
-						if (vel_max_bin >= 0) {
-							vel_max = math::min(vel_max, vel_max_bin);
-						}
-					}
+			const hrt_abstime data_age = now - _data_timestamps[i];
 
-				} else if (_obstacle_map_body_frame.distances[i] == UINT16_MAX && i == sp_index) {
-					if (!move_no_data || (move_no_data && _data_fov[i])) {
-						vel_max = 0.f;
-					}
-				}
+			if (distance < max_range) {
+				delay_distance += curr_vel_parallel * (data_age * 1e-6f);
 			}
 
-			//if the sensor field of view is zero, never allow to move (even if move_no_data=1)
-			if (num_fov_bins == 0) {
-				vel_max = 0.f;
-			}
+			const float stop_distance = math::max(0.f, distance - _min_dist_to_keep - delay_distance);
 
-			setpoint = setpoint_dir * vel_max;
-		}
+			const float max_vel = math::trajectory::computeMaxSpeedFromDistance(_param_mpc_jerk_max.get(),
+					      _param_mpc_acc_hor.get(), stop_distance, 0.f);
 
-	} else {
-		//allow no movement
-		float vel_max = 0.f;
-		setpoint = setpoint * vel_max;
+			float curr_acc_vel_constraint;
 
-		// if distance data is stale, switch to Loiter
-		if (getElapsedTime(&_last_timeout_warning) > 1_s && getElapsedTime(&_time_activated) > 1_s) {
+			if (stop_distance >= 0.f) {
```

Comment:
@MaEtUgR This is the change i added to be able to compensate velocities below the stop_distance

Reply:
But the else case is unreachable no?
`stop_distance` can only be >= 0 since you set it above as `max(0, x)` 👀 

I'd generally try to make fewer individual cases with very similar formulas and rather calculate things once and make sure that e.g. `max_vel` is zero which makes the formula the same again no?

Reply:
@MaEtUgR 
my mistake, should be fixed now.

Reply:
I see https://github.com/PX4/PX4-Autopilot/compare/3b4efab0813f57e664734440f4b0d3b3b1586150..f05f032e76e266e988500261295873d4b6391a93